source/_posts/2020-07-28-Kubernetes-controllers-A-New-Hope.md

```diff
@@ -0,0 +1,297 @@
+
+{% note %}
+Although there is a [distinction](https://stackoverflow.com/questions/47848258/kubernetes-controller-vs-kubernetes-operator) between controllers and operators, for the sake of simplicity in this post I'm going to always refer to _controller_ as the entity that implements the business logic of a CRD.
+{% endnote %}
+
+This concept works pretty well for Kubernetes, although there are still some open questions about the controller itself. In this post I'm going to introduce you a solution [Markus Thömmes](https://github.com/markusthoemmes/) and I designed that may **revolutionize** how we package, deploy and operate Kubernetes controllers. 
+
+## Kubernetes controllers today
+
+Today a Kubernetes controller consists in more or less:
+
+* A main loop that does long polling to Kubernetes to watch one or more resources
+* On every event, a function called _reconciler_ is invoked to align the cluster state to the desired state. For example: if the desired state expects a Pod to be running, but for some reason there isn't, then the _reconciler_ creates it.
+* The _reconciler_ optionally set the status of the CR, like _Ready_, _SomeError_, ...
+
+After we implement the controller, we need to configure and apply the _Custom Resource Definition_ and we need to setup a _Service Account_, that is an account to connect to our Kubernetes cluster in order to perform the operations the _reconciler_ does.
+
+During the years a lot of tooling was created, _mostly in Golang_, to implement these controllers, depending on the user needs. To mention some of them:
+
+* [Kubebuilder](https://github.com/kubernetes-sigs/kubebuilder)
+* [Operator SDK](https://sdk.operatorframework.io/)
+
+While other projects, like Knative, opt for creating and maintaining their [own tooling](https://github.com/knative/pkg/tree/master/controller) to implement controllers.
+
+These tools usually provides:
+
+* Library to improve the event handling logic: "smart" events queue that dedup events, filtering events, ...
+* A code generator component that, starting from the CRD defined in the code, scaffolds the controller code and the client code to interact with the CRD (this is because Golang doesn't have generics)
```

Comment:
```suggestion
* A code generator component that, starting from the CRD defined in the code, scaffolds the controller code and generates the client code to interact with the CRD (this is because Golang doesn't have generics)
```